cfn-resources/cloud-backup-restore-jobs/cmd/resource/resource.go

```diff
@@ -0,0 +1,341 @@
+	}
+	// API call
+	restoreJob, _, err := client.CloudProviderSnapshotRestoreJobs.Get(context.Background(), requestParameters)
+	if err != nil {
+		return handler.ProgressEvent{}, fmt.Errorf("error reading cloud provider snapshot restore job with id(project: %s, job: %s): %s", projectID, jobID, err)
+	}
+
+	currentModel.TargetClusterName = &restoreJob.TargetClusterName
+	currentModel.DeliveryType = &restoreJob.DeliveryType
+	currentModel.ExpiresAt = &restoreJob.ExpiresAt
+	currentModel.CreatedAt = &restoreJob.CreatedAt
+	currentModel.Id = &restoreJob.ID
+	currentModel.FinishedAt = &restoreJob.FinishedAt
+	currentModel.SnapshotId = &restoreJob.SnapshotID
+	currentModel.TargetProjectId = &restoreJob.TargetGroupID
+	currentModel.Timestamp = &restoreJob.Timestamp
+	currentModel.Cancelled = &restoreJob.Cancelled
+	currentModel.Expired = &restoreJob.Expired
+	currentModel.DeliveryUrl = restoreJob.DeliveryURL
+	currentModel.Links = flattenLinks(restoreJob.Links)
+	return handler.ProgressEvent{
+		OperationStatus: handler.Success,
+		Message:         "Read complete",
+		ResourceModel:   currentModel,
+	}, nil
+}
+
+// Update handles the Update event from the Cloudformation service.
+func Update(req handler.Request, prevModel *Model, currentModel *Model) (handler.ProgressEvent, error) {
+	// NO-OP
+	return handler.ProgressEvent{
+		OperationStatus: handler.Success,
+		Message:         "Update complete",
+		ResourceModel:   currentModel,
+	}, nil
+}
+
+// Delete handles the Delete event from the Cloudformation service.
+func Delete(req handler.Request, prevModel *Model, currentModel *Model) (handler.ProgressEvent, error) {
+	setup() // logger setup
+
+	_, _ = logger.Debugf("Delete snapshot restore starts for Request() currentModel:%+v", currentModel)
+	// Validate required fields in the request
+	if modelValidation := validateModel(DeleteRequiredFields, currentModel); modelValidation != nil {
+		return *modelValidation, nil
+	}
+	// Create MongoDb Atlas Client using keys
+	client, err := util.CreateMongoDBClient(*currentModel.ApiKeys.PublicKey, *currentModel.ApiKeys.PrivateKey)
+	if err != nil {
+		_, _ = logger.Warnf(errorCreateMongoClient, err)
+		return progressevents.GetFailedEventByCode(fmt.Sprintf("Failed to Create Client : %s", err.Error()),
+			cloudformation.HandlerErrorCodeInvalidRequest), nil
+	}
+
+	// Create API Request Object
+	clusterName := *currentModel.ClusterName
+	projectID := *currentModel.ProjectId
+	jobID := *currentModel.Id
+	snapshotRequest := &mongodbatlas.SnapshotReqPathParameters{
+		GroupID:     projectID,
+		ClusterName: clusterName,
+		JobID:       jobID,
+	}
+	// Check if snapshot already exist
+	if !isRestoreJobExist(currentModel) {
+		_, _ = logger.Warnf("Delete - errors reading restore with id(%s)", jobID)
+		return handler.ProgressEvent{
+			OperationStatus:  handler.Failed,
+			Message:          "Resource Not Found",
+			HandlerErrorCode: cloudformation.HandlerErrorCodeNotFound}, nil
+	}
```

Comment:
Please correct me if I am wrong but the logic for Read and Delete until "Check if snapshot already exists" seems the same to me. If so, is it possible to extract in a separate function and reuse it?

Reply:
The input field validation is different for read and delete and also log statements

Reply:
This is fixed.Please check now